rpc/dial.go

```diff
@@ -278,42 +273,45 @@ func dial(
 	return conn, cached, nil
 }
 
-func dialMulticastDNS(
-	ctx context.Context,
-	address string,
-	logger golog.Logger,
-	dOpts dialOptions,
-) (ClientConn, bool, error) {
+func lookupMDNSCandidate(ctx context.Context, address string, logger golog.Logger) (*zeroconf.ServiceEntry, error) {
 	candidates := []string{address, strings.ReplaceAll(address, ".", "-")}
-	candidateLookup := func(ctx context.Context, candidates []string) (*zeroconf.ServiceEntry, error) {
-		resolver, err := zeroconf.NewResolver(logger, zeroconf.SelectIPRecordType(zeroconf.IPv4))
-		if err != nil {
+	resolver, err := zeroconf.NewResolver(logger, zeroconf.SelectIPRecordType(zeroconf.IPv4))
+	if err != nil {
+		return nil, err
+	}
+	defer resolver.Shutdown()
+	for _, candidate := range candidates {
+		entries := make(chan *zeroconf.ServiceEntry)
+		lookupCtx, cancel := context.WithTimeout(ctx, 1*time.Second)
+		defer cancel()
+		if err := resolver.Lookup(lookupCtx, candidate, "_rpc._tcp", "local.", entries); err != nil {
+			logger.Errorw("error performing mDNS query", "error", err)
 			return nil, err
 		}
-		defer resolver.Shutdown()
-		for _, candidate := range candidates {
-			entries := make(chan *zeroconf.ServiceEntry)
-			lookupCtx, cancel := context.WithTimeout(ctx, 1*time.Second)
-			defer cancel()
-			if err := resolver.Lookup(lookupCtx, candidate, "_rpc._tcp", "local.", entries); err != nil {
-				logger.Errorw("error performing mDNS query", "error", err)
-				return nil, err
-			}
-			select {
-			case <-ctx.Done():
-				return nil, ctx.Err()
-			// entries gets closed after lookupCtx expires or there is a real entry
-			case entry := <-entries:
-				if entry != nil {
-					return entry, nil
-				}
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		// entries gets closed after lookupCtx expires or there is a real entry
+		case entry := <-entries:
+			if entry != nil {
+				return entry, nil
 			}
 		}
+	}
+	if ctx.Err() != nil {
 		return nil, ctx.Err()
 	}
-	// lookup for candidates for backward compatibility
-	entry, err := candidateLookup(ctx, candidates)
-	if err != nil || entry == nil {
+	return nil, errors.New("mDNS query failed to find a candidate")
```

Comment:
This extra guard might be unnecessary - it's not clear to me if it's practically possible to not find an entry (or find a `nil` entry) without a context error/completion.